tests/integration/records/delete-record-test.js

```diff
@@ -149,3 +150,25 @@ test("when deleted records are rolled back, they are still in their previous rec
   assert.equal(all.get('length'), 2, 'record was not removed');
   assert.equal(filtered.get('length'), 2, 'record was not removed');
 });
+
+test("Deleting an invalid newly created record should remove it from the store", function(assert) {
+  var record;
+  var store = env.store;
+  run(function() {
+    record = store.createRecord('person', { name: 'pablobm' });
+    // Invalidate the record to put it in the `root.loaded.created.invalid` state
+    store.recordWasInvalid(record._internalModel, { name: 'error' });
+  });
+
+  // Preconditions
+  assert.equal(get(record, 'currentState.stateName'), 'root.loaded.created.invalid',
+                 'records should start in the created.invalid state');
+  assert.equal(get(store.peekAll('person'), 'length'), 1, 'The new person should be in the store');
+
+  run(function() {
+    record.deleteRecord();
+  });
+
+  assert.equal(get(record, 'currentState.stateName'), 'root.deleted.saved');
+  assert.equal(get(store.peekAll('person'), 'length'), 0, 'The new person should be removed from the store');
+});
```

Comment:
Can you add a test case for `destroyRecord` too? I think this fix should also work for #3688.